kernel/msg_q.c

```diff
@@ -187,6 +187,100 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 	return result;
 }
 
+int z_impl_k_msgq_put_urgent(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+{
+	__ASSERT(!arch_is_in_isr() || K_TIMEOUT_EQ(timeout, K_NO_WAIT), "");
+
+	struct k_thread *pending_thread;
+	k_spinlock_key_t key;
+	int result;
+
+	key = k_spin_lock(&msgq->lock);
+
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);
+
+	if (msgq->used_msgs < msgq->max_msgs) {
+		/* message queue isn't full */
+		pending_thread = z_unpend_first_thread(&msgq->wait_q);
+		if (unlikely(pending_thread != NULL)) {
+			SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, 0);
+
+			/* give message to waiting thread */
+			(void)memcpy(pending_thread->base.swap_data, data,
+			       msgq->msg_size);
```

Comment:
Not loving the duplication of the implementation code.  Ideally all this should be shared.  In particular maintenance-quality-issues are at their most important in code surrounding memory copies into/out-of caller pointers (which can cross security boundaries if CONFIG_USERSPACE=y!).  Might take some refactoring to make that happen, obviously.

Reply:
I understand and agree, but at first I really didn't want to touch in the already existing `k_msgq_put` function. So should I do something like this, then?

```c
int _copy_to_queue ( ... , bool should_copy_to_back) {
  // shared implementation goes here
}

int z_impl_k_msgq_put( ... ) {
  return _copy_to_queue( ... , true);
}

int z_impl_k_msgq_put_urgent( ... ) {
  return _copy_to_queue( ..., , false);
}
```

Reply:
FWIW I have done just that to avoid code duplication.